src/devices/bus/hp_hil/hlekbd.h

```diff
@@ -6,64 +6,92 @@
 #pragma once
 
 #include "hp_hil.h"
+#include "hlebase.h"
 #include "machine/keyboard.h"
 
 
-namespace bus { namespace hp_hil {
+namespace bus {
+	namespace hp_hil {
 
-class hle_device_base
-	: public device_t
-	, public device_hp_hil_interface
-	, protected device_matrix_keyboard_interface<15U>
+class hle_hp_ipc_device
+		: public hle_device_base
+		, protected device_matrix_keyboard_interface<15U>
 {
 public:
-	virtual ~hle_device_base() override;
-
-	static constexpr feature_type imperfect_features() { return feature::KEYBOARD; }
-
-protected:
-	// constructor/destructor
-	hle_device_base(machine_config const &mconfig, device_type type, char const *tag, device_t *owner, uint32_t clock);
+	hle_hp_ipc_device(machine_config const &mconfig, char const *tag, device_t *owner, uint32_t clock);
 
-	// device overrides
-	virtual void device_start() override;
+	virtual ioport_constructor device_input_ports() const override;
 	virtual void device_reset() override;
 
 	// device_matrix_keyboard_interface overrides
 	virtual void key_make(uint8_t row, uint8_t column) override;
 	virtual void key_break(uint8_t row, uint8_t column) override;
-
-	// device_hp_hil_interface overrides
-	virtual bool hil_write(uint16_t *data) override;
-
+	virtual int hil_poll() override;
+	virtual void hil_idd() override;
 private:
-	void transmit_byte(uint8_t byte);
-
 	util::fifo<uint8_t, 8> m_fifo;
+	void transmit_byte(uint8_t byte);
 };
 
-
-class hle_hp_ipc_device : public hle_device_base
+class hle_hp_itf_device
+		: public hle_device_base
+		, protected device_matrix_keyboard_interface<15U>
 {
 public:
-	hle_hp_ipc_device(machine_config const &mconfig, char const *tag, device_t *owner, uint32_t clock);
-
+	hle_hp_itf_device(machine_config const &mconfig, char const *tag, device_t *owner, uint32_t clock);
+	virtual void device_reset() override;
 	virtual ioport_constructor device_input_ports() const override;
+
+	// device_matrix_keyboard_interface overrides
+	virtual void key_make(uint8_t row, uint8_t column) override;
+	virtual void key_break(uint8_t row, uint8_t column) override;
+	virtual int hil_poll() override;
+	virtual void hil_idd() override;
+private:
+	util::fifo<uint8_t, 8> m_fifo;
```

Comment:
The trouble with adding `util::fifo` is it doesn't support save states (it can't because it's in `src/lib/util` so it can't be aware of things in `src/emu`).